File: ScriptsToProcess/NEW BATCH TO PROCESS/Vertical Projection.py
import Rhino.Geometry as rg

def project_point_vertically(pt, plane):
    """
    Projects a point vertically (along world Z-axis) onto a given plane.
    
    Args:
        pt (rg.Point3d): The input point.
        plane (rg.Plane): The target plane.

    Returns:
        tuple: (bool, projected_point, message)
    """
    # Compute the signed distance from the point to the plane
    dist = plane.DistanceTo(pt)

    # Create a vertical line from the point (World Z-axis)
    vertical_dir = rg.Vector3d(0, 0, 1)  # Global Z-axis
    vertical_line = rg.Line(pt - vertical_dir * 10000, pt + vertical_dir * 10000)  # Long enough line

    # Find intersection of the vertical line with the plane
    success, parameter = rg.Intersect.Intersection.LinePlane(vertical_line, plane)

    if success:
        projected_pt = vertical_line.PointAt(parameter)  # Get the actual intersection point

        # Determine if the point is on, above, or below the plane
        if abs(dist) < 1e-6:
            message = "Point is on the plane\nNo projection needed"
        elif dist > 0:
            message = "Point was above the plane\nProjected downward"
        else:
            message = "Point was below the plane\nProjected upward"

        return abs(dist) < 1e-6, projected_pt, message

    return False, pt, "No valid projection\nCheck input values"  # Fallback case

# Grasshopper Inputs:
# pt: Input point (Type Hint: Point3d)
# plane: Input plane (Type Hint: Plane)

is_on_plane, projected_pt, msg = project_point_vertically(pt, plane)

# Grasshopper Outputs:
isContained = is_on_plane  # Boolean if the point is on the plane
ProjectedPoint = projected_pt  # The actual projected Point3d
Message = msg  # Multi-line message output

# Set Grasshopper component metadata
if 'ghenv' in globals():
    ghenv.Component.Name = "Vertical Projection"
    ghenv.Component.NickName = "VertProj"
    ghenv.Component.Message = msg

File: Components/VerticalProjection.cs
using System;
using Grasshopper.Kernel;
using Rhino.Geometry;

namespace Enzyme.Components
{
    public class VerticalProjectionComponent : GH_Component
    {
        public VerticalProjectionComponent()
          : base("Vertical Projection", "VertProj",
              "Projects a point vertically (along world Z-axis) onto a given plane.",
              "Enzyme", "Pattern")
        {
        }

        protected override void RegisterInputParams(GH_InputParamManager pManager)
        {
            pManager.AddPointParameter("Point", "pt", "Input point", GH_ParamAccess.item);
            pManager.AddPlaneParameter("Plane", "plane", "Input plane", GH_ParamAccess.item);
        }

        protected override void RegisterOutputParams(GH_OutputParamManager pManager)
        {
            pManager.AddBooleanParameter("Is Contained", "isContained", "Boolean if the point is on the plane", GH_ParamAccess.item);
            pManager.AddPointParameter("Projected Point", "ProjectedPoint", "The actual projected Point3d", GH_ParamAccess.item);
            pManager.AddTextParameter("Message", "Message", "Multi-line message output", GH_ParamAccess.item);
        }

        protected override void SolveInstance(IGH_DataAccess DA)
        {
            Point3d pt = Point3d.Unset;
            Plane plane = Plane.Unset;

            if (!DA.GetData(0, ref pt)) return;
            if (!DA.GetData(1, ref plane)) return;

            // Compute the signed distance from the point to the plane
            double dist = plane.DistanceTo(pt);

            // Create a vertical line from the point (World Z-axis)
            Vector3d verticalDir = new Vector3d(0, 0, 1);
            Line verticalLine = new Line(pt - verticalDir * 10000.0, pt + verticalDir * 10000.0);

            // Find intersection of the vertical line with the plane
            bool success = Rhino.Geometry.Intersect.Intersection.LinePlane(verticalLine, plane, out double parameter);

            bool isContained = false;
            Point3d projectedPt = pt;
            string message = "";

            if (success)
            {
                projectedPt = verticalLine.PointAt(parameter);

                if (Math.Abs(dist) < 1e-6)
                {
                    message = "Point is on the plane\nNo projection needed";
                    isContained = true;
                }
                else if (dist > 0)
                {
                    message = "Point was above the plane\nProjected downward";
                }
                else
                {
                    message = "Point was below the plane\nProjected upward";
                }
            }
            else
            {
                message = "No valid projection\nCheck input values";
            }

            // HUD Message
            this.Message = message;

            DA.SetData(0, isContained);
            DA.SetData(1, projectedPt);
            DA.SetData(2, message);
        }

        protected override System.Drawing.Bitmap Icon
        {
            get
            {
                return IconLoader.Load("VerticalProjection.png");
            }
        }

        public override Guid ComponentGuid
        {
            get { return new Guid("B95F5651-7667-4638-A3BA-F0121F247920"); }
        }
    }
}
